apispec/core.py

```diff
@@ -213,35 +227,61 @@ def normalize_path(path):
             if path and 'basePath' in self.options:
                 pattern = '^{0}'.format(re.escape(self.options['basePath']))
                 path = re.sub(pattern, '', path)
-
             return path
 
         if isinstance(path, Path):
             path.path = normalize_path(path.path)
+            if operations:
+                path.operations.update(operations)
         else:
             path = Path(path=normalize_path(path),
                         operations=operations,
-                        openapi_version=self.openapi_version.vstring)
-        # Execute plugins' helpers
+                        openapi_version=self.openapi_version)
+
+        # Execute path helpers
+        for plugin in self.plugins:
+            try:
+                ret = plugin.path_helper(path=path, operations=path.operations, **kwargs)
+            except (NotImplementedError, TypeError):
+                continue
+            if isinstance(ret, Path):
+                ret.path = normalize_path(ret.path)
+                path.update(ret)
+        # Deprecated interface
         for func in self._path_helpers:
             try:
                 ret = func(
-                    self, path=path, operations=operations, **kwargs
+                    self, path=path, operations=path.operations, **kwargs
                 )
             except TypeError:
                 continue
             if isinstance(ret, Path):
                 ret.path = normalize_path(ret.path)
                 path.update(ret)
-                operations = ret.operations
-        if operations:
-            for func in self._operation_helpers:
-                func(self, path=path, operations=operations, **kwargs)
-
         if not path.path:
             raise APISpecError('Path template is not specified')
 
-        # Process response helpers for any path operations defined.
+        # Execute operation helpers
+        for plugin in self.plugins:
+            try:
+                plugin.operation_helper(path=path, operations=path.operations, **kwargs)
+            except NotImplementedError:
```

Comment:
It's possible for this to swallow `NotImplementedError` errors that aren't raised directly by plugin methods. I suppose `BasePlugin` could raise a plugin-specific exception, e.g. `PluginMethodNotImplementedError` to avoid this. Might not be absolutely necessary, though.

Reply:
Yes, I'm not sure sure it is that bad. I'm more uncomfortable with the `TypeError` caught above. I already experienced issues because it was hiding real problems in my code. `NotImplementedError` is more specific. But I don't mind changing this if you wish.

Reply:
Yeah, I'm also concerned about TypeErrors being swallowed. I don't recall why that exception-handling was added. Would have to dig more into it.

Reply:
Perhaps was it made to avoid breaking on helper functions that would not have **kwargs in their signature?

I could remove it while I'm at it (no test broken on this branch if I do).

Reply:
Ah, that seems likely. Yeah, go ahead and remove it.

Reply:
OK, I removed `TypeError` catching for the new plugin interface and left it on the legacy helpers.

I also introduced `PluginMethodNotImplementedError`.